In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
df = pd.read_csv("./data/230907_basic_table_for_analysis_cleaned.csv", index_col=0, low_memory=False)

df.head()

/Users/johannastarkl/opt/anaconda3/lib/python3.8/site-packages/numpy/lib/arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


num_veh  place  catu  grav  sexe  an_nais  trajet  locp actp  \
Num_Acc                                                                     
201900000001     B01    2.0     2     4     2   2002.0     0.0  -1.0   -1   
201900000001     B01    1.0     1     4     2   1993.0     5.0  -1.0   -1   
201900000001     A01    1.0     1     1     1   1959.0     0.0  -1.0   -1   
201900000002     A01    1.0     1     4     2   1994.0     0.0  -1.0   -1   
201900000003     A01    1.0     1     1     1   1996.0     0.0  -1.0    0   

              etatp  ...  obs  obsm  choc  manv        date  is_holiday  \
Num_Acc              ...                                                  
201900000001   -1.0  ...  0.0   2.0   5.0  23.0  2019-11-30       False   
201900000001   -1.0  ...  0.0   2.0   5.0  23.0  2019-11-30       False   
201900000001   -1.0  ...  1.0   0.0   3.0  11.0  2019-11-30       False   
201900000002   -1.0  ...  4.0   0.0   1.0   0.0  2019-11-30       False   
201900000003   -1.0  ...  0.0   2.0   1.0   2.0  2019-11-28       False   

              holiday  secu_used  secu_avail fatal  
Num_Acc                                             
201900000001      NaN          3           1     0  
201900000001      NaN          3           1     0  
201900000001      NaN          3           1     0  
201900000002      NaN          3           1     0  
201900000003      NaN          3           1     0  

[5 rows x 46 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2509596 entries, 201900000001 to 201800057783
Data columns (total 46 columns):
 #   Column        Dtype  
---  ------        -----  
 0   num_veh       object 
 1   place         float64
 2   catu          int64  
 3   grav          int64  
 4   sexe          int64  
 5   an_nais       float64
 6   trajet        float64
 7   locp          float64
 8   actp          object 
 9   etatp         float64
 10  an            int64  
 11  mois          int64  
 12  jour          int64  
 13  hrmn          object 
 14  lum           int64  
 15  agg           int64  
 16  int           int64  
 17  atm           float64
 18  col           float64
 19  adr           object 
 20  lat           float64
 21  long          float64
 22  dep           int64  
 23  metropolitan  float64
 24  catr          float64
 25  circ          float64
 26  nbv           float64
 27  vosp          float64
 28  prof          float64
 29  plan          float64
 30  su

In [4]:
df['date'] = pd.to_datetime(df.date, format='%Y-%m-%d')
df['year'] = df['date'].dt.year.astype(int)

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2509596 entries, 201900000001 to 201800057783
Data columns (total 47 columns):
 #   Column        Dtype         
---  ------        -----         
 0   num_veh       object        
 1   place         float64       
 2   catu          int64         
 3   grav          int64         
 4   sexe          int64         
 5   an_nais       float64       
 6   trajet        float64       
 7   locp          float64       
 8   actp          object        
 9   etatp         float64       
 10  an            int64         
 11  mois          int64         
 12  jour          int64         
 13  hrmn          object        
 14  lum           int64         
 15  agg           int64         
 16  int           int64         
 17  atm           float64       
 18  col           float64       
 19  adr           object        
 20  lat           float64       
 21  long          float64       
 22  dep           int64         
 23  metropolitan  fl

In [5]:
df.shape

(2509596, 47)

In [6]:
#drop columns not relevant for model

df.drop(['adr', 'lat', 'long', 'holiday','date'], axis=1, inplace=True)


In [7]:
df.shape
df.head()

num_veh  place  catu  grav  sexe  an_nais  trajet  locp actp  \
Num_Acc                                                                     
201900000001     B01    2.0     2     4     2   2002.0     0.0  -1.0   -1   
201900000001     B01    1.0     1     4     2   1993.0     5.0  -1.0   -1   
201900000001     A01    1.0     1     1     1   1959.0     0.0  -1.0   -1   
201900000002     A01    1.0     1     4     2   1994.0     0.0  -1.0   -1   
201900000003     A01    1.0     1     1     1   1996.0     0.0  -1.0    0   

              etatp  ...  occutc  obs  obsm choc  manv  is_holiday  secu_used  \
Num_Acc              ...                                                        
201900000001   -1.0  ...    -1.0  0.0   2.0  5.0  23.0       False          3   
201900000001   -1.0  ...    -1.0  0.0   2.0  5.0  23.0       False          3   
201900000001   -1.0  ...    -1.0  1.0   0.0  3.0  11.0       False          3   
201900000002   -1.0  ...    -1.0  4.0   0.0  1.0   0.0       False          3   
201900000003   -1.0  ...    -1.0  0.0   2.0  1.0   2.0       False          3   

              secu_avail  fatal  year  
Num_Acc                                
201900000001           1      0  2019  
201900000001           1      0  2019  
201900000001           1      0  2019  
201900000002           1      0  2019  
201900000003           1      0  2019  

[5 rows x 42 columns]

In [8]:
#if cell before doesn't work

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

df['num_veh'] = le.fit_transform(df['num_veh'].astype(str))
df['actp'] = le.fit_transform(df['actp'].astype(str))
df['hrmn'] = le.fit_transform(df['hrmn'].astype(str))


In [9]:
df.head()

num_veh  place  catu  grav  sexe  an_nais  trajet  locp  actp  \
Num_Acc                                                                       
201900000001       14    2.0     2     4     2   2002.0     0.0  -1.0     0   
201900000001       14    1.0     1     4     2   1993.0     5.0  -1.0     0   
201900000001        0    1.0     1     1     1   1959.0     0.0  -1.0     0   
201900000002        0    1.0     1     4     2   1994.0     0.0  -1.0     0   
201900000003        0    1.0     1     1     1   1996.0     0.0  -1.0     1   

              etatp  ...  occutc  obs  obsm  choc  manv  is_holiday  \
Num_Acc              ...                                              
201900000001   -1.0  ...    -1.0  0.0   2.0   5.0  23.0       False   
201900000001   -1.0  ...    -1.0  0.0   2.0   5.0  23.0       False   
201900000001   -1.0  ...    -1.0  1.0   0.0   3.0  11.0       False   
201900000002   -1.0  ...    -1.0  4.0   0.0   1.0   0.0       False   
201900000003   -1.0  ...    -1.0  0.0   2.0   1.0   2.0       False   

              secu_used  secu_avail  fatal  year  
Num_Acc                                           
201900000001          3           1      0  2019  
201900000001          3           1      0  2019  
201900000001          3           1      0  2019  
201900000002          3           1      0  2019  
201900000003          3           1      0  2019  

[5 rows x 42 columns]

In [10]:
#separating into test and training data: 

#for training data keep years 2015-2019 
train = df[df['year'] < 2020]

#for test data keep years 2020-2021
test = df[df['year'] >= 2020]



In [11]:
#define X_train, X_test, y_train, y_test

X_train = train.drop(['fatal'], axis=1)
y_train = train['fatal']

X_test = test.drop(['fatal'], axis=1)
y_test = test['fatal']

In [12]:
#DecisionTree Classifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score, accuracy_score, classification_report
from sklearn.model_selection import GridSearchCV


# Tuning the parameters to find the best decision tree model - start params chosen randomly
model = DecisionTreeClassifier(criterion = 'gini', random_state=50)

param_dist = {'criterion': ['gini', 'entropy'],
              'max_depth': [2, 3, 4], 
              'min_samples_leaf' : [2,3,4]}

cv_dt = GridSearchCV(model, param_grid=param_dist, n_jobs=-1)

cv_dt.fit(X_train, y_train)
print('Best Parameters using grid search: \n', cv_dt.best_params_)



Best Parameters using grid search: 
 {'criterion': 'gini', 'max_depth': 2, 'min_samples_leaf': 2}


In [14]:
# Tuning the parameters to find the best decision tree model - start params chosen randomly
model = DecisionTreeClassifier(random_state=50)

param_dist = {'criterion': ['gini', 'entropy'],
              'max_depth': [2,5,6,7], 
              'min_samples_leaf' : [2,5,6,7]}

cv_dt = GridSearchCV(model, param_grid=param_dist, n_jobs=-1)

cv_dt.fit(X_train, y_train)
print('Best Parameters using grid search: \n', cv_dt.best_params_)


Best Parameters using grid search: 
 {'criterion': 'gini', 'max_depth': 2, 'min_samples_leaf': 2}


In [17]:
# Training with the best parameters

tree = DecisionTreeClassifier(criterion='gini', random_state = 50, max_depth=2, min_samples_leaf=2)
tree.fit(X_train, y_train)
y_pred_gini  = tree.predict(X_test)

print('Accuracy Score: ',accuracy_score(y_test, y_pred_gini))


Accuracy Score:  0.9997440797106395


In [23]:
pd.crosstab(y_test, y_pred_gini, rownames=['Reality'], colnames=['Prediction'])

Prediction       0     1
Reality                 
0           228389     0
1               60  5999

In [22]:
# classification_report

print('Classification Report')
print(classification_report(y_test, y_pred_gini))



Classification Report
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    228389
           1       1.00      0.99      1.00      6059

    accuracy                           1.00    234448
   macro avg       1.00      1.00      1.00    234448
weighted avg       1.00      1.00      1.00    234448



In [18]:
#Training set accuracy

y_pred_train_gini = tree.predict(X_train)

y_pred_train_gini

print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train_gini)))

Training-set accuracy score: 1.0000


In [20]:
# print the scores on training and test set

print('Training set score: {:.4f}'.format(tree.score(X_train, y_train)))

print('Test set score: {:.4f}'.format(tree.score(X_test, y_test)))

Training set score: 1.0000
Test set score: 0.9997


Here, the training-set accuracy score is 1.0000 while the test-set accuracy to be 0.9997. These two values are quite comparable. So, there is no sign of overfitting.